### ISSUE TITLE
Handle long sessions

### ISSUE THREAD

**Issue by Rich-Harris** [sveltejs/kit#87]
In [Second-guessing the modern web](https://macwright.com/2020/05/10/spa-fatigue.html), Tom MacWright makes several good points (I responded [here](https://dev.to/richharris/in-defense-of-the-modern-web-2nia)).

The biggest, to me, is this one:

> User sessions are surprisingly long: someone might have your website open in a tab for weeks at a time. I’ve seen it happen. So if they open the ‘about page’, keep the tab open for a week, and then request the ‘home page’, then the home page that they request is dictated by *the index bundle that they downloaded last week*. This is a deeply weird and under-discussed situation.

I wrote:

> It's an excellent point that isn't really being addressed, though (as Tom acknowledges) it's really just exacerbating a problem that was always there. I think there are solutions to it — we can iterate on the 'index bundle' approach, we could include the site version in a cookie and use that to show actionable feedback if there's a mismatch — but we do need to spend time on it.

It would be great if there was some mechanism for solving (or at least alleviating) this problem. I think the 'include the site version in a cookie' approach is the most promising, but I haven't thought through any of the details.

**Comment by benmccann** [comment]
For posterity, here was the corresponding Sapper issue: https://github.com/sveltejs/sapper/issues/389

**Comment by oodavid** [comment]
I could see users falling into various camps as to how to approach this issue.

For us, we have an "your app is out of date" message. Which is great for "dashboard" type sites, but not so good for blogs etc.

<img width="591" alt="Screenshot 2021-03-30 at 09 48 26" src="https://user-images.githubusercontent.com/46879/112961329-2e415c00-913d-11eb-9566-b5e8df30ebdf.png">

For Svelte, I've seen this handled with a `navigator.serviceWorker` `statechange` event. Example code:

https://github.com/tobinbradley/Mecklenburg-County-GeoPortal/blob/c22980c0915e48be18e292cb5f4d9cc221e41a09/src/js/registerServiceWorker.js

I'd love to see this approach well documented in the `$service-worker` module.

My biggest concern with this issue isn't the the inner-workings, but rather the learning-curve. Most users will have to experience the issue before they start to look for a solution. For many it will be a heisenbug. I'm tempted to request that the default behaviour should be a _little_ intrusive.

`Flutter` has a [debugShowCheckedModeBanner](https://api.flutter.dev/flutter/material/MaterialApp/debugShowCheckedModeBanner.html) flag, which enables a bunch of debug features. One of the side-effects is that a little "debug" ribbon is visible on all views. This is part of the onboarding experience. It helps early users dig a little deeper into the intention of their work. This could be useful for this particular issue.

**Comment by CaptainCodeman** [comment]
I've used a web-socket subscription (using Firebase RTDB) to a release version or timestamp to trigger the service-worker update check.

That can then force a refresh or prompt the user, depending on the app itself and what they may be doing with it.

Updates that don't require all the clients to update can use the regular service-worker check (when the app is restarted) and the release version or timestamp only updated when you need to force everyone to update to the latest ASAP.

**Comment by sserdyuk** [comment]
This problem exists for any web app that uses chunking with timestamps. For example, we have the same issue with Angular & Webpack. I haven't really solved it yet but I thought an elegant way of solving it would be a service worker that receive a list of chunks from the bundler and precaches them in the background. This way the app is cached as one consistent build and continues working even if a new build has been uploaded. Just an idea.

**Comment by Nick-Mazuk** [comment]
Would it help if the page is refreshed in the background? You could probably use the Page Visibility API to detect if a user is viewing the site. If not (e.g., they're on another tab) and there's a new version, force a reload. Then, the refresh is hidden to the user.

https://developer.mozilla.org/en-US/docs/Web/API/Page_Visibility_API

Don't think this will completely solve the issue (e.g., antony's checkout example above), but it should help limit the need for a "refresh the page" toast.

**Comment by WaltzingPenguin** [comment]
I'm worried that nearly all suggestions in this thread presuppose an active server.  I hope the use case of writing docs in SvelteKit, exporting the project to HTML, and throwing the result up on Surge/Netlify/etc won't get second class treatment.

Simply reloading the page on navigation if the required files cannot be imported covers a lot of ground.  The default code splitting is done on page boundaries, so this is when the error occurs most frequently.  The different URL means we should be able to reload without impact; the only times this isn't true would also break if someone opened that link in a new tab.  No polling or version numbers are required.

The big cases where this approach fails:
* Dynamic imports within the same page - This can be mitigated by making it easier to prefetch these chunks.  If it's already loaded in to the browser cache, it doesn't matter if it disappears from the server.
* API incompatibility - I don't think a generic solution will ever work for this.  The needs for applications vary a lot here.

**Comment by tgecho** [comment]
The way I’ve usually handled this is to upload the pre-generated chunks to a separate static file service (usually S3). With filenames containing the MD5 of their contents, the chunks from multiple versions can coexist as long as needed, and older unchanged chunks can remain cached downstream.

It’s slightly messy operationally, but for my purposes the simplicity and robustness is well worth it. Unfortunately, I’m not sure if this approach can be abstracted in a portable way.

**Comment by sserdyuk** [comment]
I think keeping old versions of chunks around is a working but highly inconvenient solution because this sort of compound versioning isn't natively available from (any that I know of) cloud providers.

I think precaching chunks is the most elegant way. As long as they have the same or longer (than the html page itself) cache retention set, it will guarantee that full code tree is cached in the browser. If we know the chunk list from the bundler, we can inject `<link rel="prefetch" href="foo.js">` into the html header to achieve that effect.

**Comment by phelbas** [comment]
Not seen mentioned here is making sure whatever solution for this works also for mobile progressive web apps. I've had trouble even intercepting errors and putting out a message that invokes window.location.reload(true). If browser reload is not working or the message says "reload the page" and they are in a PWA, it is confusing.  With service workers and so on things get really difficult really quickly. See: https://deanhume.com/displaying-a-new-version-available-progressive-web-app/

**Comment by mquandalle** [comment]
For reference I was able to port my Sapper workaround into SvelteKit without much change (cf. https://github.com/sveltejs/sapper/issues/389#issuecomment-778787004). The idea remain to fetch every 10 seconds a static file `mysite.com/_version` containing a timestamp of the last deployment, and if the timestamp has changed reload the app on the next page navigation — details are in the sapper thread.

The only change was to put the `_version` file in the `static/` directory using

```json
"scripts": {
  "postinstall": "date '+%s%N' > static/_version"
}
```

**Comment by Rich-Harris** [comment]
Closed via #3412 (thanks @pz-mxu!) I've opened #3666 and #3667 as potential follow-ups.
